### ISSUE TITLE
curl_easy_perform throwing an error on the Simple example

### ISSUE THREAD

**Issue by brunoss** [bagder/curl#628]
I thought I should bring [my question from stackoverflow](http://stackoverflow.com/questions/35088146/curl-throws-error-on-curl-easy-perform) to here.

It has been some time that worked with c/c++ and I had some issues linking the curl library to my project on the beginning, but now I think I am doing it fine, I am able to build my project.
I also tried to compile the code with different options up to no avail. 

Calling `curl_easy_perform` throws an error - invalid heap pointer. With debugging your source code I found that the error occurs in `Curl_dedotdotify`.

What can be the causes for this? Thanks for your help.

**Comment by bagder** [comment]
Interesting. Can you you show us exact where in that function it happens?

**Comment by brunoss** [comment]
@bagder Like I told in stackoverflow, in the last free. http://snag.gy/l9FAy.jpg

**Comment by bagder** [comment]
And what is the 'input' string that is passed in to the Curl_dedotdotify() pointing to at the start of the function, before it crashes?

**Comment by brunoss** [comment]
It is a pointer to '/'

**Comment by bagder** [comment]
We test exactly that case in our unit test 1395, I don't see how that can fail for you! That function doesn't do much with that simple input, can you see it do anything wrong if you single-step through it?

**Comment by brunoss** [comment]
If i step through the free the program hangs.

**Comment by bagder** [comment]
Sure, but that is just an indication that something has messed up the heap already, it isn't necessarily this particular function. And I really don't think it is this function that did it. And this is with an unchanged libcurl example source code?

**Comment by brunoss** [comment]
Yeah I thought about that too, usually when it comes to this kind of errors, they occur earlier than the moment when they appear. Yeah, like I told you I am running the simple sample from the site. Can this problem be related to some kind of proxy?

**Comment by bagder** [comment]
No, this cannot be caused by a proxy or a firewall.

**Comment by jay** [comment]
TL;DR: This issue looks a lot like a CRT mismatch. Check that you are not mixing debug and release static libraries.

---

The C/C++ command line parameters you posted look as though they are from a debug configuration. I notice you have specified the include directory to your release build of libcurl, maybe you did the same for the lib directory? If so that's a problem.

I can reproduce what you described in VS2012.5 if in the Debug configuration I link libcurl's release lib instead of its debug lib. When I compile that way I see this, which is usually a bad sign:
~~~
1>LINK : warning LNK4098: defaultlib 'MSVCRT' conflicts with use of other libs; use /NODEFAULTLIB:library
~~~

What is happening when I do that is the libcurl static lib links to CRT msvcr110.dll (Release) and the application links to msvcr110d.dll (Debug) and that combination leads to problems because the memory management is different.

When you are building a static library in Visual Studio you should do so for both Debug and Release configurations if you use both configurations. Try these instructions:

---

Open a _VS2012 x86 Native Tools Command Prompt_
Switch to your curl winbuild dir, run these commands:
~~~
nmake /f Makefile.vc mode=static
nmake /f Makefile.vc mode=static DEBUG=yes
~~~
(If you need x64 versions as well you will need to open a _VS2012 x64 Cross Tools Command Prompt_ and repeat the above commands with `MACHINE=x64` appended and modify the directions below for x64 instead of x86.)

Now in your curl dir you should have
~~~
builds\libcurl-vc-x86-debug-static-ipv6-sspi-winssl
builds\libcurl-vc-x86-release-static-ipv6-sspi-winssl
~~~

For Debug and Release configurations in the Visual Studio project do the following from the properties page:

**C/C++ > General > Additional Include directories**
Release: Prepend `C:\<yourcurldir>\builds\libcurl-vc-x86-release-static-ipv6-sspi-winssl\include;`
Debug: Prepend `C:\<yourcurldir>\builds\libcurl-vc-x86-debug-static-ipv6-sspi-winssl\include;`

**C/C++ > Precompiled Headers > Precompiled Header**
Release & Debug: `Not Using Precompiled Headers`
(Unless you are using pch in which case ignore this)

**C/C++ > Preprocessor > Preprocessor Definitions**
Release & Debug: Prepend `CURL_STATICLIB;`

**Linker > General > Additional Library directories**
Release: Prepend `C:\<yourcurldir>\builds\libcurl-vc-x86-release-static-ipv6-sspi-winssl\lib;`
Debug: Prepend `C:\<yourcurldir>\builds\libcurl-vc-x86-debug-static-ipv6-sspi-winssl\lib;`

**Linker > Input > Additional Dependencies**
Release: Prepend `libcurl_a.lib;`
Debug: Prepend `libcurl_a_debug.lib;`

---

If this is the answer to your SO question please answer it yourself so others who read your question will know the answer. Note for the future if you are unsure whether you are experiencing a bug try one of [the mailing lists](http://curl.haxx.se/mail/) for a wider audience.

**Comment by brunoss** [comment]
Thank you very much @jay.  Very thoughtful and with a very complete explanation. Sorry for the trouble.

**Comment by patrikhson** [comment]
Hi,

I've been looking for some useful stuff and found an interesting site with useful advices, please take a look <http://stentiprake.justthetipin.org/xycrxdyx>

My best to you, paf@frobbit.se
